main.tf

```diff
@@ -0,0 +1,409 @@
+        storage_class = "STANDARD_IA"
+      }
+
+      transition {
+        days          = 60
+        storage_class = "GLACIER"
+      }
+
+      expiration {
+        days = 90
+      }
+    }
+  }
+
+
+}
+
+locals {
+  lambda_role_arns = compact([
+    module.slack_notifier.role_arn,
+    module.slack_event_listener.role_arn,
+    module.ldap_query_lambda.role_arn,
+    module.dynamodb_cleanup.role_arn
+  ])
+}
+
+resource "aws_s3_bucket_policy" "artifacts" {
+  bucket = aws_s3_bucket.artifacts.id
+  policy = <<POLICY
+{
+  "Version": "2012-10-17",
+  "Id": "lambda_access",
+  "Statement": [
+    {
+        "Effect": "Allow",
+        "Principal": {
+            "AWS": ${jsonencode(local.lambda_role_arns)}
+        },
+        "Action": [
+            "s3:GetObject",
+            "s3:PutObject",
+            "s3:DeleteObject"
+        ],
+        "Resource": "${aws_s3_bucket.artifacts.arn}/*"
+    }
+  ]
+}
+  POLICY
+}
+
+locals {
+  lambda_function_arns = compact([
+    module.ldap_query_lambda.function_arn,
+    module.slack_notifier.function_arn,
+    module.dynamodb_cleanup.function_arn
+  ])
+}
+
+# step function
+data "aws_iam_policy_document" "sfn" {
+  statement {
+    actions = [
+      "lambda:InvokeFunction"
+    ]
+    resources = local.lambda_function_arns
+  }
+}
+
+resource "aws_iam_policy" "sfn" {
+  name        = "${var.project_name}-sfn"
+  description = "Policy used by the ${var.project_name} Step Function"
+  policy      = "${data.aws_iam_policy_document.sfn.json}"
+}
+
+data "aws_iam_policy_document" "trust" {
+  statement {
+    actions = ["sts:AssumeRole"]
+    principals {
+      type        = "Service"
+      identifiers = ["states.amazonaws.com"]
+    }
+  }
+}
+
+resource "aws_iam_role" "sfn" {
+  name = "${var.project_name}-sfn"
+
+  assume_role_policy = data.aws_iam_policy_document.trust.json
+}
+
+resource "aws_iam_policy_attachment" "sfn" {
+  name       = "${var.project_name}-sfn"
+  roles      = ["${aws_iam_role.sfn.name}"]
+  policy_arn = "${aws_iam_policy.sfn.arn}"
+}
+
+resource "aws_sfn_activity" "account_deactivation_approval" {
+  name = "account_deactivation_approval"
+}
+
+locals {
+  dynamodb_maintenance_sfn_content = var.create_dynamodb_cleanup ? templatefile(
+    "${path.module}/templates/dynamodb_cleanup_task.tpl",
+    {
+      function_arn = module.dynamodb_cleanup.function_arn
+  }) : ""
+  # list of parallel states to run as part of the ldap object clean up
+  # **Note: this list must start with a comma
+  additional_cleanup_tasks_internal = var.create_dynamodb_cleanup ? ",${local.dynamodb_maintenance_sfn_content}" : ""
+  additional_cleanup_tasks          = var.additional_cleanup_tasks == "" ? local.additional_cleanup_tasks_internal : "${local.additional_cleanup_tasks_internal}, ${var.additional_cleanup_tasks}"
+
+}
+
+resource "aws_sfn_state_machine" "ldap_maintenance" {
+  name     = var.project_name
+  role_arn = "${aws_iam_role.sfn.arn}"
+
+  definition = <<EOF
+{
```

Comment:
Think we should move the definition to a template_file

Reply:
moved